File: C#/RefactoringWorkshop/Test/Coupler/MessageChain/ArticleWrapperConverterTest.cs
using System;
using System.Collections.Generic;
using NUnit.Framework;
using RefactoringWorkshop.Coupler.MessageChain;
using Rhino.Mocks;

namespace Test.Coupler.MessageChain
{
    public class ArticleWrapperConverterTest
    {
        [Test]
        public void should_convert_article_response_wrapper_to_article_wrapper() {
            var articleResponseWrapper = new ArticleResponseWrapper();

            var articleResponse = MockRepository.GenerateStub<ArticleResponse>();
            articleResponse.Articles = new List<Article> {GetArticle()};

            var subscriptionPreferenceWrapper = MockRepository.GenerateStub<SubscriptionPreferenceWrapper>();
            subscriptionPreferenceWrapper.EntityCode = "code";
            subscriptionPreferenceWrapper.EntityName = "company";

            var subscriptionWrapper = MockRepository.GenerateStub<SubscriptionWrapper>();
            subscriptionWrapper.SubscriptionPreferences = new List<SubscriptionPreferenceWrapper> {subscriptionPreferenceWrapper};

            articleResponseWrapper.ArticleResponse = articleResponse;
            articleResponseWrapper.CompanySubscription = subscriptionWrapper;

            var converter = new ArticleWrapperConverter();
            var articleWrapper = converter.Convert(articleResponseWrapper);

            Assert.AreEqual(articleWrapper.Companies[0], "company");
            Assert.AreEqual(articleWrapper.EntityCodes[0], "CODE");
            Assert.AreEqual(articleWrapper.ArticleRef, "articleRef");
        }

        [Test]
        public void should_return_empty_article_wrapper_when_there_are_no_articles() {
            var articleResponse = new ArticleResponse();
            ArticleResponseWrapper articleResponseWrapper =
                new ArticleResponseWrapper {ArticleResponse = articleResponse};
            var converter = new ArticleWrapperConverter();
            var articleWrapper = converter.Convert(articleResponseWrapper);

            Assert.True(articleWrapper.Companies.Count == 0);
            Assert.True(articleWrapper.EntityCodes.Count == 0);
            Assert.IsNull(articleWrapper.ArticleRef);
        }

        [Test]
        public void should_return_article_ref_when_company_subscription_is_not_available() {
            var articleResponse = new ArticleResponse();
            var articles = new List<Article> {GetArticle()};
            articleResponse.Articles = articles;
            var articleResponseWrapper = new ArticleResponseWrapper {ArticleResponse = articleResponse};
            var converter = new ArticleWrapperConverter();

            var articleWrapper = converter.Convert(articleResponseWrapper);

            Assert.True(articleWrapper.Companies.Count == 0);
            Assert.True(articleWrapper.EntityCodes.Count == 0);
            Assert.AreEqual(articleWrapper.ArticleRef, "articleRef");
        }

        [Test]
        public void should_return_article_ref_when_metadata_codes_is_not_available() {
            var articleResponse = new ArticleResponse();
            var articles = new List<Article>();
            var article = GetArticle();
            article.MetadataCodes = null;
            articles.Add(article);
            articleResponse.Articles = articles;
            var articleResponseWrapper = new ArticleResponseWrapper {ArticleResponse = articleResponse};
            var converter = new ArticleWrapperConverter();

            var articleWrapper = converter.Convert(articleResponseWrapper);

            Assert.True(articleWrapper.Companies.Count == 0);
            Assert.True(articleWrapper.EntityCodes.Count == 0);
            Assert.AreEqual(articleWrapper.ArticleRef, "articleRef");
        }

        private Article GetArticle()
        {
            Article article = new Article();
            MetadataCodes metadataCodes = new MetadataCodes();
            metadataCodes.AddCompanyCode("code", "company");
            article.MetadataCodes = metadataCodes;
            article.ArticleRef = "articleRef";
            return article;
        }
    }
}

File: Java/RefactoringWorkshop/src/test/java/coupler/messagechain/ArticleWrapperConverterTest.java
package coupler.messagechain;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.runners.MockitoJUnitRunner;

import java.util.ArrayList;
import java.util.List;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertThat;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.isNull;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class ArticleWrapperConverterTest {

    @Test public void
    should_convert_article_response_wrapper_to_article_wrapper() {
        ArticleResponseWrapper articleResponseWrapper = new ArticleResponseWrapper();

        ArticleResponse articleResponse = mock(ArticleResponse.class);
        SubscriptionWrapper subscriptionWrapper = mock(SubscriptionWrapper.class);

        SubscriptionPreferenceWrapper subscriptionPreferenceWrapper = mock(SubscriptionPreferenceWrapper.class);
        when(subscriptionPreferenceWrapper.getEntityCode()).thenReturn("code");
        when(subscriptionPreferenceWrapper.getEntityName()).thenReturn("company");
        List<SubscriptionPreferenceWrapper> subscriptionPreferenceWrappers = new ArrayList<>();
        subscriptionPreferenceWrappers.add(subscriptionPreferenceWrapper);
        when(subscriptionWrapper.getSubscriptionPreferences()).thenReturn(subscriptionPreferenceWrappers);

        List<Article> articles = new ArrayList<>();
        articles.add(getArticle());

        when(articleResponse.getArticles()).thenReturn(articles);

        articleResponseWrapper.setArticleResponse(articleResponse);
        articleResponseWrapper.setCompanySubscription(subscriptionWrapper);
        ArticleWrapperConverter converter = new ArticleWrapperConverter();
        ArticleWrapper articleWrapper = converter.convert(articleResponseWrapper);

        assertThat(articleWrapper.getCompanies().get(0), is("company"));
        assertThat(articleWrapper.getEntityCodes().get(0), is("CODE"));
        assertThat(articleWrapper.getArticleRef(), is("articleRef"));
    }

    @Test public void
    should_return_empty_article_wrapper_when_there_are_no_articles() {
        ArticleResponse articleResponse = new ArticleResponse();
        ArticleResponseWrapper articleResponseWrapper = new ArticleResponseWrapper();
        articleResponseWrapper.setArticleResponse(articleResponse);
        ArticleWrapperConverter converter = new ArticleWrapperConverter();
        ArticleWrapper articleWrapper = converter.convert(articleResponseWrapper);

        assertTrue(articleWrapper.getCompanies().isEmpty());
        assertTrue(articleWrapper.getEntityCodes().isEmpty());
        assertNull(articleWrapper.getArticleRef());
    }

    @Test public void
    should_return_article_ref_when_company_subscription_is_not_available() {
        ArticleResponse articleResponse = new ArticleResponse();
        ArrayList<Article> articles = new ArrayList<>();
        articles.add(getArticle());
        articleResponse.setArticles(articles);
        ArticleResponseWrapper articleResponseWrapper = new ArticleResponseWrapper();
        articleResponseWrapper.setArticleResponse(articleResponse);
        ArticleWrapperConverter converter = new ArticleWrapperConverter();

        ArticleWrapper articleWrapper = converter.convert(articleResponseWrapper);

        assertTrue(articleWrapper.getCompanies().isEmpty());
        assertTrue(articleWrapper.getEntityCodes().isEmpty());
        assertThat(articleWrapper.getArticleRef(), is("articleRef"));
    }

    @Test public void
    should_return_article_ref_when_metadata_codes_is_not_available() {
        ArticleResponse articleResponse = new ArticleResponse();
        ArrayList<Article> articles = new ArrayList<>();
        Article article = getArticle();
        article.setMetadataCodes(null);
        articles.add(article);
        articleResponse.setArticles(articles);
        ArticleResponseWrapper articleResponseWrapper = new ArticleResponseWrapper();
        articleResponseWrapper.setArticleResponse(articleResponse);
        ArticleWrapperConverter converter = new ArticleWrapperConverter();

        ArticleWrapper articleWrapper = converter.convert(articleResponseWrapper);

        assertTrue(articleWrapper.getCompanies().isEmpty());
        assertTrue(articleWrapper.getEntityCodes().isEmpty());
        assertThat(articleWrapper.getArticleRef(), is("articleRef"));

    }
    private Article getArticle() {
        Article article = new Article();
        MetadataCodes metadataCodes = new MetadataCodes();
        metadataCodes.addCompanyCode("code", "company");
        article.setMetadataCodes(metadataCodes);
        article.setArticleRef("articleRef");
        return article;
    }
}